LOOP-1200: Persist Alerts to CoreData

rickpasetto opened the pull request:
The general idea was to follow the pattern in `DeviceLog`.  Since the design is similar, I decided to call it the `DeviceAlertLog`, with the following schema:

Name | Type | Optional? | Description 
------ | ----- | ---------- |  ------------
acknowledgedTimestamp | Date | YES | Recorded if the alert is acknowledged
backgroundContent | String | YES | rendered as a JSON blob (see `DeviceAlert.Content`)
foregroundContent | String | YES | rendered as a JSON blob (see `DeviceAlert.Content`)
identifier | String | NO | unique alert _type_ per device (includes device identifier)
isCritical | Boolean | NO | if either foreground or background is true, this is true (Not sure if this is best...)
issuedTimestamp | Date | NO | Recorded when the alert is issued *this, along with identifier, is the unique key*
modificationCounter | Int64 | NO | monotonically increasing number
retractedTimestamp | Date | YES | Recorded if the alert is retracted
sound | String | YES | rendered as a JSON blob (see `DeviceAlert.Sound`)
trigger | String | YES | rendered as a JSON blob (see `DeviceAlert.Trigger`)~

One thing that was required here was to squash the two API functions for `removePendingAlert` and `removeDeliveredAlert` into `retractAlert`.  There was, as yet, no reason for these functions to be separate, and since we now needed to record when an alert was "retracted", it made sense to combine them.  This caused (once again) a fairly small but extensive API change. 

See LW PR: https://github.com/tidepool-org/LoopWorkspace/pull/98

[LOOP-1200](https://tidepool.atlassian.net/browse/LOOP-1200)

rickpasetto committed:
LOOP-1200: Persist DeviceAlerts to CoreData

Also: Squash removePendingAlert and removeDeliveredAlert into retractAlert

ps2 commented:
There are going to be some non-device-manager level alerts (storage space running low, bluetooth disabled, etc) that should probably also end up in this storage, be tracked with acknowledgement, and be shown in any history displays.  So maybe this one shouldn't be called "DeviceAlert". It's also a little different than a "Log" in that entries are updated. Following the pattern of other stores in loop, maybe we call it "AlertStore"?

rickpasetto commented:
@ps2 Is it just the name you're concerned about?  I'm totally fine renaming these things:
`DeviceAlertLog.xcdatamodeld` -> `AlertStore.xcdatamodeld`
`DeviceAlertLogEntry.swift`-> `AlertStoreEntry.swift`
`PersistentDeviceAlertLog.swift` -> `PersistentAlertStore.swift`

Was that what you were thinking?  How about the CoreData schema...was that ok with you? (one thing in particular was the separate `managerIdentifier` and `alertIdentifier`... if there are non-device-manager alerts, perhaps we should just call it `identifier` and assume that they'd be combined?)

ps2 commented:
Haven't done a full review yet just wanted to get that feedback to you asap since it seems central and related to requirements.

rickpasetto commented:
Copy that, no problem :)

ps2 commented:
re other names: 
`PersistentAlertStore` -> `AlertStore`? 
`AlertStoreEntry` -> `Alert` or `StoredAlert`?,

rickpasetto commented:
> re other names:
> `PersistentAlertStore` -> `AlertStore`?

`AlertStore` sounds good to me

> AlertStoreEntry -> Alert or StoredAlert?,

I prefer `StoredAlert` if you're ok with that

rickpasetto committed:
PR Feedback: Renaming

rickpasetto commented:
Closing this PR until I've done some re-work per @ps2 's request

rickpasetto committed:
checkpoint

rickpasetto committed:
Checkpoint: back to codegen'd CoreData object (seems to work)

rickpasetto committed:
Remove bits about expiration until we know how/when we want to do it.

rickpasetto committed:
Shuffled some code around a little.

rickpasetto committed:
oops these changes weren't staged.

rickpasetto commented:
@ps2 I've updated this PR with a new data model, per our discussion.  Please take a look while I write some unit tests.

rickpasetto committed:
Add unit tests, remove query support (for another PR)

darinkrauss commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -0,0 +1,149 @@
+//
+//  PersistentDeviceAlertLog.swift
```
[nit] AlertStore.swift?

darinkrauss commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -0,0 +1,149 @@
+        let storeDescription = NSPersistentStoreDescription()
+        if let storageFileURL = storageFileURL {
+            storeDescription.url = storageFileURL
+            print("AlertStore: \(storageFileURL)")
```
[nit] print?

rickpasetto replied:
👍 I'll delete it.  FWIW This is handy to see where the sqlite database is located so you can open it up to see what's inside.

darinkrauss commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -0,0 +1,149 @@
+        }
+        persistentContainer = NSPersistentContainer(name: "AlertStore")
+        persistentContainer.persistentStoreDescriptions = [storeDescription]
+        persistentContainer.loadPersistentStores { description, error in
```
[nit] `description` -> `_`?

darinkrauss commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -0,0 +1,149 @@
+
+extension AlertStore {
+    
+    public func recordIssued(alert: DeviceAlert, at timestamp: Date = Date(), completion: ((Result<Void, Error>) -> Void)? = nil) {
```
[nit] I'm getting some naming confusion here. This is the `AlertStore` that deals with `StoredAlerts` with a parameter name `alert`, but a type of `DeviceAlert`?

rickpasetto replied:
yeah...all of this used to be called `DeviceAlert*`, but I see that now my renaming is making things a little more confusing. I don't know where or when we'll have "non-device alerts" as Pete suggested earlier in this PR, but I'd imagine we'd add functions here to record those as well.  It may involve a refactor where there might be a generic `Alert` (protocol?), of which `DeviceAlert` is one.  At any rate, until we start building those "non-device alerts" I will leave this as-is.

darinkrauss replied:
What about alerts that just originate from Loop directly?

rickpasetto replied:
That's what I mean by "non-device alerts".  They don't use the new Alert Facility yet.

darinkrauss commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -0,0 +1,149 @@
+    public func recordIssued(alert: DeviceAlert, at timestamp: Date = Date(), completion: ((Result<Void, Error>) -> Void)? = nil) {
+        dataAccessQueue.async {
+            self.managedObjectContext.perform {
+                _ = StoredAlert(from: alert, context: self.managedObjectContext, issuedTimestamp: timestamp)
```
[nit] The rest of our Core Data models use words with "Date" not "Timestamp".

rickpasetto replied:
Ah...good to know 👍 .  I'll change all of the `*Timestamp` fields to `*Date`.

darinkrauss commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -0,0 +1,149 @@
+                    self.log.default("Recorded alert: %{public}@", alert.identifier.value)
+                    completion?(.success)
+                } catch {
+                    self.log.error("Could not store alert: %{public}@", String(describing: error))
```
[nit] Consider adding the alert that could not be stored to this error?

darinkrauss commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -0,0 +1,149 @@
+    
+    private func lookupLatest(identifier: DeviceAlert.Identifier, completion: @escaping (Result<StoredAlert?, Error>) -> Void) {
+        managedObjectContext.perform {
+            let predicate = NSPredicate(format: "identifier = %@", identifier.value)
```
[nit] I prefer `==` over `=` so it reads more like code, but whatever your preference is.

rickpasetto replied:
you mean in the predicate?  Sure, that sounds like a good rule. 👍

darinkrauss commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -0,0 +1,149 @@
+            do {
+                let fetchRequest: NSFetchRequest<StoredAlert> = StoredAlert.fetchRequest()
+                fetchRequest.predicate = predicate
+                fetchRequest.sortDescriptors = [ NSSortDescriptor(key: "modificationCounter", ascending: true) ]
```
Might be better to use descending sort with a limit of 1? This way you only fetch one record no matter how many are in the database.

rickpasetto replied:
Good idea. 👍

darinkrauss commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -0,0 +1,149 @@
+                    switch $0 {
+                    case .success(let object):
+                        if let object = object {
+                            block(object)
```
Any need to have the block return an error?

Alternatively, consider passing the *optional* object to the block and having it return an error so the block can decide if a record not found is an error or not.

rickpasetto replied:
These are good suggestions, if this was a more public function.  This is a private implementation detail, though, of `recordRetraction` and `recordAcknowledgement`, and my goal was to minimize code duplication in those functions.  As such, the optional check and return (or "completion") of an error would also be duplicated.  
I'm not strongly opposed to your suggestion, I just don't really see a need for it.  Thoughts?

darinkrauss replied:
Works for me.

darinkrauss commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -0,0 +1,149 @@
+                                self.log.default("Recorded alert: %{public}@", identifier.value)
+                                completion?(.success)
+                            } catch {
+                                self.log.error("Could not store alert: %{public}@", String(describing: error))
```
[nit] Again, log actual alert here, too?

darinkrauss commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -0,0 +1,149 @@
+extension AlertStore {
+
+    // At the moment, this is only used for unit testing
+    func fetch(identifier: DeviceAlert.Identifier, completion: @escaping (Result<[StoredAlert], Error>) -> Void) {
```
[nit] `fileprivate` so outsiders don't use fetch independently? (I'm dealing with this in the other stores while trying to track down all uses of the store. Restricting direct access to store through limited number of intentional public APIs is a good thing.)

darinkrauss replied:
Unless of course, outsiders should be using `fetch`?

rickpasetto replied:
See the comment above the function.  I wish there was a way to restrict access only to unit tests, but the best we can do to make it `internal` (the default) and not `public`.  I will add an explicit `internal` keyword to make this even clearer.

darinkrauss replied:
If possible, consider moving this extension into the test file itself? If you use an `@testable` import, I *think* you should be able to access to the internal properties (maybe?).

darinkrauss commented on Loop/Managers/Alerts/DeviceAlertManager.swift:
```diff
@@ -85,19 +89,24 @@ extension DeviceAlertManager: DeviceAlertManagerResponder {
-        handlers.forEach { $0.removeDeliveredAlert(identifier: identifier) }
+    
+    private func replayAlert(_ alert: DeviceAlert) {
+        handlers.forEach { $0.issueAlert(alert) }
```
[nit] Any need to check to see if the alert is still not in a retracted or acknowledged state?

rickpasetto replied:
Excellent question!  This (private) function assumes that retraction actually removes the alert from wherever it was issued ... in this case, UserNotificationCenter, which is still where we replay alerts.

darinkrauss commented on Loop/Managers/Alerts/InAppModalDeviceAlertPresenter.swift:
```diff
@@ -50,23 +50,18 @@ public class InAppModalDeviceAlertPresenter: DeviceAlertPresenter {
-    }
-    
+        
     // For tests only
```
[nit] Is this comment still true? I see `removeDeliveredAlert` used in `retractAlert`.

darinkrauss commented on Loop/Managers/Alerts/InAppModalDeviceAlertPresenter.swift:
```diff
@@ -100,6 +95,7 @@ extension InAppModalDeviceAlertPresenter {
                 return
             }
             self.playSound(for: alert)
+            print("Alert: \(try! alert.encodeToString())")
```
[nit] `print`?

darinkrauss commented on Loop/Managers/Alerts/StoredAlert.swift:
```diff
@@ -0,0 +1,48 @@
+            identifier = deviceAlert.identifier.value
+            // Encode as JSON strings
+            let encoder = StoredAlert.encoder
+            trigger = try encoder.encodeToStringIfPresent(deviceAlert.trigger)
```
[nit] Why encode these to `string` and not binary `Data`? I suppose the same question for why using the `JSONEncoder`?

rickpasetto replied:
For small amounts of data, I find it handy to encode in something human-readable, so I can double-check data by looking at the database, and it is easier to write unit test.  But this is totally up for debate; changing this later will be hard (because it would require a migration).  WDYT?

darinkrauss replied:
I suppose it depends upon how much JSON would be generated for this. Either way, up to you.

rickpasetto replied:
Cool.  Again, if there's a general rule that "blobs" of data should be binary, not JSON or otherwise, I'm ok to change it.  For now I will leave it.

darinkrauss commented on Loop/Managers/Alerts/StoredAlert.swift:
```diff
@@ -0,0 +1,48 @@
+            foregroundContent = try encoder.encodeToStringIfPresent(deviceAlert.foregroundContent)
+            backgroundContent = try encoder.encodeToStringIfPresent(deviceAlert.backgroundContent)
+            isCritical = deviceAlert.foregroundContent?.isCritical ?? false || deviceAlert.backgroundContent?.isCritical ?? false
+            modificationCounter = 0
```
[nit] The other Core Data models do not explicitly initialize `modificationCounter` in code (because I didn't want folks thinking that it was something that should be manually modified. I believe they all configure the property in the model file to default to 0.

darinkrauss replied:
Up to you, though.

rickpasetto replied:
👍 Nice.  I'd rather do it that way.

darinkrauss commented on Loop/Managers/Alerts/StoredAlert.swift:
```diff
@@ -0,0 +1,48 @@
+    func encodeToStringIfPresent<T>(_ encodable: T?) throws -> String? where T: Encodable {
+        guard let encodable = encodable else { return nil }
+        let data = try self.encode(encodable)
+        return String(data: data, encoding: .utf8)
```
[nit] In theory, `String(data:encoding:)` can return `nil` if the string cannot be encoded, which would be a hidden error. Not likely given encoding is UTF-8, but...

rickpasetto replied:
Good catch

darinkrauss reviewed (approved):
A pile of nits, but nothing blocking from what I can see. LGTM! 🌊

rickpasetto committed:
PR Feedback, more unit tests

ps2 commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -0,0 +1,150 @@
+
+extension AlertStore {
+    
+    public func recordIssued(alert: DeviceAlert, at date: Date = Date(), completion: ((Result<Void, Error>) -> Void)? = nil) {
```
I saw Darin pointed out that this is taking `DeviceAlert`.  It could accept a separate protocol or struct called `NewAlert` (we have a similar pattern with NewPumpEvent).  If it's a protocol, then DeviceAlert could just conform to it. The "New" in the name is not perfect, as we use the same structs in the pump apis to update existing records, and I'd think you'd do the same here. But it's just an idea.

rickpasetto replied:
Honestly, I had thought of this when I first made `DeviceAlert`.  You may recall that my initial name for that struct was `UserAlert` (meaning: an alert for the _User_).  There really is nothing "device-ish" about it...the only thing coming close is the `managerIdentifier` (which, conveniently and somewhat intentionally, also doesn't have the word _device_ in it). 

I have nothing against protocols (love them! I sprinkle them on my breakfast!), but in this case, I'm not sure I see the value.  They are good when abstracting complex objects (managers, delegates, etc.) but in this case it is really just a package of immutable information, an ideal choice for a `struct` IMHO. (In fact, `DeviceAlert` has no external dependencies...it's just nested structs of primitive types).

I think I'd much rather change `DeviceAlert` to just `Alert`, and then these names will all match.  The downside of course is that this name matches a `SwiftUI` object as well as (I think) an object in the DexcomSDK.  However, the Swift compiler should protect us in that case (e.g. `SwiftUI.Alert` vs `LoopKit.Alert`). 

WDYT?  If you agree, I'll make a follow-up PR

ps2 replied:
I think that's fine; if there are no special distinguishing factors between alerts issued by devices and general system alerts. Only thing I can think of is tracking where it came from.  Your central alert manager needs to send acknowledgements for alerts from devices back to their device managers, and you don't need to do that for system alerts. If fulfilling that responsibility entails changing the alert structure (adding a device manager id, for example), then there is some difference.

ps2 replied:
I'm guessing your plan is just to keep that device manager id in all the contexts (app level, storage, and devicemanager interface), and I think that's ok. I tend to think having specific structs for specific contexts is nice because it helps keep concerns from spreading, and provides helps keeps the boundaries explicit. But the cost is extra structs, and conversions between them. Sometimes that's nice in terms of organizing and keeping each context clean, and sometimes it's just extra work.

rickpasetto replied:
Yes, my plan is to keep `managerIdentifier`, not necessarily meaning "device manager".  I hadn't thought of `managerIdentifier` as anything other than a namespace scoping (i.e. "this alert came from that manager") -- but that might be because of my only recently working in the codebase. 
When we start doing "system" alerts, I imagine they'll need to be scoped as well.  One might come from the "loop algorithm" area (which may be in LoopKit), another from the "system level" (e.g. disk space low) area (which may be in Loop), and yet others from (backend) Services (which are separate plugins, though I can't come up with what alert we would want from them). 
The `Identifier` structure is meant to be exactly the rendezvous mechanism for alert acknowledgement, so we know where to send what acknowledgment.
Perhaps the name `managerIdentifier` is the place we're getting tripped up?
At any rate, I understand you concern, and I can see how to go the other route should we choose it.  
(I'll re-request your approval for all of this.  If you still disagree, happy to chat more about it.)

ps2 replied:
The AlertManager at the LoopLevel is within the context of Loop itself, and it would make complete sense for it to be the "handler" for an acknowledgment, and would not need further routing.  But again, this is a small thing, and I think it's acceptable having a global alert with a device identifier in it even for alerts that don't need further routing.

ps2 commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -0,0 +1,150 @@
+    }
+    
+    public func recordAcknowledgement(of identifier: DeviceAlert.Identifier, at date: Date = Date(),
+                                      completion: ((Result<Void, Error>) -> Void)? = nil) {
```
The `(Result<Void, Error>)` is interesting. We have typically just used `Error?` in this case.  They are very close, type-wise. Seems like there are tradeoffs both ways. It's somewhat more explicit to use the Result type, but you lose a little syntactic shortcut like `if let error = error` in handling the response (It becomes`if case .error(let error) = error`).

rickpasetto replied:
I prefer `Result` because:
1) it forces callers to think about the error case.  It is much easier to ignore the `Error?` passed in, but the wrapper `Result` forces you to unwrap it to get the "actual" result.
2) I guess IMHO `if let error = error` and `if case .error(let error) = error` are both weird.  That's why (you'll notice) I prefer a `switch` statement.  
3) Using the `Void` version (`Result<Void, Error>`) gets you used to the pattern, so the code looks the same as the non-`Void` versions.  (It's handy to have the 
```swift
extension Result where Success == Void {
    static var success: Result {
        return Result.success(Void())
    }
}
```
code below, too.)

WDYT?

ps2 replied:
Yeah, I think it's fine. It seems a little less idiomatic, but I don't feel strongly about it. Just saw it was a new pattern and thought it was interesting.
I think you can ignore an Optional (which is actually very similar to a Result) just as easily, and can unwrap it the same, but Optional has a little more built in syntax support.

rickpasetto replied:
I had been using my own `Result` type for a while before Swift officially added it. I took it as a cue that this was a better way to go, and emboldened me to eschew the `Error?` route.

I'm not sure I follow your counterargument about ignoring an Optional... that's what my (1) argument above...?

As far as syntax support... I'm not sure if there are plans for nicer syntax support in Swift.

ps2 replied:
I was just saying I think it's just as easy to ignore an `Error?` type result as it is a `Result<Void, Error>` type, so I don't think argument #1 applies. I love using `Result` when there is an actual success type (not Void) that callers may need.  And as far as syntax niceties, I'm just talking about `Optionals`, even though they're a simple enum like `Result`, having some syntatic shortcuts, like `if let`, optional chaining, unwrapping, etc. 

I'm not opposed to `Result<Void, Error>` at all, and may use it if I feel the explicitness helps a particular case. It seems a little more heavy than `Error?`, but sometimes being more explicit is nice.

ps2 commented on Loop/Managers/Alerts/DeviceAlertManager.swift:
```diff
@@ -44,18 +44,22 @@ public final class DeviceAlertManager {
         self.userNotificationCenter = userNotificationCenter
         self.fileManager = fileManager
+        let documentsDirectory = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first!
+        alertStore = AlertStore(storageFileURL: documentsDirectory.appendingPathComponent("AlertStore.sqlite"))
```
Might want to put this into it's own directory, to contain and allow easy cleanup of the group of files that this creates (the write ahead log, shm file, etc.)

rickpasetto replied:
Good idea!  I'll just use a subdirectory called `AlertStore`, unless you want to stuff other storage in that directory too...

ps2 reviewed (approved):
LGTM. Had some suggestions, but nothing blocking.

rickpasetto committed:
PR feedback

rickpasetto committed:
Merge 'dev' into 'rick/LOOP-1200-alert-persistence'

rickpasetto committed:
Add a log line for when an alert is not found to update

ps2 commented on Loop/Managers/Alerts/AlertStore.xcdatamodeld/AlertStore.xcdatamodel/contents:
```diff
@@ -0,0 +1,18 @@
+        <attribute name="modificationCounter" attributeType="Integer 64" defaultValueString="0" usesScalarValueType="YES"/>
+        <attribute name="retractedDate" optional="YES" attributeType="Date" usesScalarValueType="NO"/>
+        <attribute name="sound" optional="YES" attributeType="String"/>
+        <attribute name="trigger" optional="YES" attributeType="String"/>
```
I was looking for the delay attribute for future scheduled alerts, and couldn't find it, and then realized it must be encoded in the trigger JSON.  Do you see any current need, or a likely future need for querying events based on when they would be issued? If so, maybe consider encoding the Trigger fields flatly in this object.

rickpasetto replied:
Yeah, after talking with Paul today about how they may want to display the Notifications, I think we may have display policy around future vs. immediate alerts.  I'm going to change this now, to avoid migration later.

ps2 reviewed (approved):
LGTM! Just a question and possible change if you see a need.

darinkrauss reviewed (approved):
LGTM! ✂️

rickpasetto committed:
Merge branch 'dev' into rick/LOOP-1200-alert-persistence

rickpasetto committed:
PR Feedback: "flatten" Trigger so that it is queryable

Stored trigger type as an Int16, trigger interval as an optional Double (NSNumber).

rickpasetto commented:
@ps2 I took your suggestion to "flatten" the `trigger` storage in the database.  Could you please have one more look before I start merging?  Thanks.

ps2 reviewed (approved):
LGTM!

rickpasetto committed:
Add query using modification counter (#92)

* Add query using modification counter

* PR Feedback

* LOOP-1200: Dump Alert database into Issue Report (#93)

* LOOP-1200: Dump Alert database into Issue Report

Added query by date, plus unit tests.

* PR Feedback, plus some rework/refactor

This takes @darin's suggestion, with some modifications :).  Now, the Filter is part of the Anchor, which makes it so anchoring continues to apply the filter as appropriate.

rickpasetto committed:
LOOP-1200: alert persistence followups (#94)

* Add query using modification counter

* PR Feedback

* LOOP-1200: Dump Alert database into Issue Report

Added query by date, plus unit tests.

* PR Feedback, plus some rework/refactor

This takes @darin's suggestion, with some modifications :).  Now, the Filter is part of the Anchor, which makes it so anchoring continues to apply the filter as appropriate.

* LOOP-1200: Split alertIdentifier and managerIdentifier

After talking with Paul, we determined that we will need to be able to either filter or group alerts by device, so it would be better to keep managerIdentifier separate now so it can be used in data queries.
